x/evm/keeper/context_stack.go

```diff
@@ -0,0 +1,78 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type cachedContext struct {
+	ctx    sdk.Context
+	commit func()
+}
+
+// ContextStack manages the initial context and a stack of cached contexts,
+// to support the `StateDB.Snapshot` and `StateDB.RevertToSnapshot` methods.
+type ContextStack struct {
+	// Context of the initial state before transaction execution.
+	// It's the context used by `StateDB.CommitedState`.
+	initialCtx     sdk.Context
+	cachedContexts []cachedContext
+}
+
+// CurrentContext returns the top context of cached stack,
+// if the stack is empty, returns the initial context.
+func (k *ContextStack) CurrentContext() sdk.Context {
+	l := len(k.cachedContexts)
+	if l == 0 {
+		return k.initialCtx
+	}
+	return k.cachedContexts[l-1].ctx
+}
+
+// Reset sets the initial context and clear the cached context stack,
+// panic if the cached context stack is not empty, to avoid unintential state loss.
```

Comment:
I think the panic is on `RevertAll` instead

Reply:
good catch, the comment is out dated, removed now.